packages/framework/ai-collab/src/explicit-strategy/agentEditReducer.ts

```diff
@@ -0,0 +1,570 @@
+
+import { assert } from "@fluidframework/core-utils/internal";
+import { isFluidHandle } from "@fluidframework/runtime-utils";
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+import {
+	Tree,
+	NodeKind,
+	type ImplicitAllowedTypes,
+	type TreeArrayNode,
+	type TreeNode,
+	type TreeNodeSchema,
+	type TreeView,
+	type SimpleNodeSchema,
+	FieldKind,
+	FieldSchema,
+	isTreeNodeSchemaClass,
+	normalizeAllowedTypes,
+	normalizeFieldSchema,
+	type ImplicitFieldSchema,
+	booleanSchema,
+	handleSchema,
+	nullSchema,
+	numberSchema,
+	stringSchema,
+} from "@fluidframework/tree/internal";
+
+import {
+	type TreeEdit,
+	type ObjectTarget,
+	type Selection,
+	type Range,
+	type ObjectPlace,
+	type ArrayPlace,
+	type TreeEditObject,
+	type TreeEditValue,
+	typeField,
+} from "./agentEditTypes.js";
+import type { IdGenerator } from "./idGenerator.js";
+import type { JsonValue } from "./jsonTypes.js";
+import { toDecoratedJson } from "./promptGeneration.js";
+import { fail } from "./utils.js";
+
+function populateDefaults(
+	json: JsonValue,
+	definitionMap: ReadonlyMap<string, SimpleNodeSchema>,
+): void {
+	if (typeof json === "object") {
+		if (json === null) {
+			return;
+		}
+		if (Array.isArray(json)) {
+			for (const element of json) {
+				populateDefaults(element, definitionMap);
+			}
+		} else {
+			assert(
+				typeof json[typeField] === "string",
+				"The typeField must be present in new JSON content",
+			);
+			const nodeSchema = definitionMap.get(json[typeField]);
+			assert(nodeSchema?.kind === NodeKind.Object, "Expected object schema");
+		}
+	}
+}
+
+function getSchemaIdentifier(content: TreeEditValue): string | undefined {
+	switch (typeof content) {
+		case "boolean":
+			return booleanSchema.identifier;
+		case "number":
+			return numberSchema.identifier;
+		case "string":
+			return stringSchema.identifier;
+		case "object": {
+			if (content === null) {
+				return nullSchema.identifier;
+			}
+			if (Array.isArray(content)) {
+				// TODO: Support arrays for setRoot
+				throw new UsageError("Arrays are not currently supported in this context");
+			}
+			if (isFluidHandle(content)) {
+				return handleSchema.identifier;
+			}
+			return content[typeField];
+		}
+		default:
+			throw new UsageError("Unsupported content type");
+	}
+}
+
+function isConstructable(schema: TreeNodeSchema): boolean {
+	switch (schema.identifier) {
+		case booleanSchema.identifier:
+		case numberSchema.identifier:
+		case stringSchema.identifier:
+		case nullSchema.identifier:
+		case handleSchema.identifier:
+			return false;
+		default:
+			return isTreeNodeSchemaClass(schema);
+	}
+}
+
+function contentWithIds(content: TreeNode, idGenerator: IdGenerator): TreeEditObject {
+	return JSON.parse(toDecoratedJson(idGenerator, content)) as TreeEditObject;
+}
+
+/**
+ * Manages applying the various types of {@link TreeEdit}'s to a a given {@link TreeNode}.
+ */
+export function applyAgentEdit<TSchema extends ImplicitFieldSchema>(
```

Comment:
This won't work for trees with no `TreeNode`s in them (e.g. primitive as the root, polymorphic type at the root where one type is a primitive, etc.). This should probably either allow undefined for the scope (can we rename `treeNode` to `nodeScope` or something more descriptive?) or take a field for the scope which we could type as (TreeNode * fieldKey) | TreeView.

Reply:
or TreeNode | TreeView, that would be fine too.

Reply:
After thinking about this a little bit more, I feel like the setRoot operation is just poorly defined in general if you allow subtree scoping. For example, what happens if the user passes in an object scope that is replaced by a set root operation halfway through a series of edits made by the agent? Any subsequent usages of the scope node will result in a thown exception, as it will be a removed subtree. Perhaps what we really need here is to establish the scope as the parent/key pair for whatever node scope the user passed in? Or perhaps we simply get rid of the set root operation entirely. @Josmithr @noencke thoughts?

Reply:
Yea, can we handle all insertion/new content with an input of "parent node + field key"? That covers children of objects, children of arrays, and the root (in which case the parent is undefined a.k.a. null for GPT). Then there's no such thing as setRoot, there's just a special case of insert.

Reply:
Ah, but if they pass in array node + index (say, 4) as the parent scope, and the agent removes the root of the scope that's passed in...the array shifts and it gets a new scope, namely the entry at old index 5 (or undefined if index 4 was the end of the array)?? Arrays at the root make everything so weird.

Reply:
I agree that this is weird. FYI I have removed the setRoot operation as well as the ability to pass the TreeView in commit fb37d5500e19de5dd5a5664513592c997115e358, We can revisit adding it back later.